clippy_utils/src/source.rs

```diff
@@ -55,158 +93,563 @@ impl<'sm> HasSourceMap<'sm> for &LateContext<'sm> {
-    ) -> Option<Range<BytePos>> {
-        map_range(sm.source_map(), self.into_range(), f)
+        f: impl for<'a> FnOnce(&'a SpanEditCx<'sm>, FileRange) -> Option<T>,
+    ) -> Option<T::Output>
+    where
+        Self: SpanLike,
+    {
+        if let Some((scx, range)) = SpanEditCx::for_local(sm.source_map(), self.data())
+            && let Some(ranges) = f(&scx, range.clone())
+        {
+            ranges.dbg_check(&scx, range);
+            Some(ranges.into_spans(&scx))
+        } else {
+            None
+        }
     }
 
-    /// Extends the range to include all preceding whitespace characters.
+    /// Walks this span up the macro call chain to the target context. Returns `None` if the target
+    /// context cannot be found.
     ///
-    /// The range will not be expanded if it would cross a line boundary, the line the range would
-    /// be extended to ends with a line comment and the text after the range contains a
-    /// non-whitespace character on the same line. e.g.
+    /// Since both early and late lint passes see the expanded code the only way to detect macro
+    /// expanded code is via a span's `SyntaxContext`. This function can be used to get e.g. the
+    /// span of function argument as typed at the caller rather than the span of macro generated
+    /// code that makes up the argument's value.
     ///
-    /// ```ignore
-    /// ( // Some comment
-    /// foo)
+    /// Given the following code:
+    ///
+    /// ```rust,ignore
+    /// macro_rules! m1 { ($e:expr) => { f1($e) }; }
+    /// macro_rules! m2 { ($e:expr) => { f2(m1!($e)) }; }
+    /// f3(m2!(0))
     /// ```
     ///
-    /// When the range points to `foo`, suggesting to remove the range after it's been extended will
-    /// cause the `)` to be placed inside the line comment as `( // Some comment)`.
-    fn with_leading_whitespace<'sm>(self, sm: impl HasSourceMap<'sm>) -> Range<BytePos> {
-        with_leading_whitespace(sm.source_map(), self.into_range())
+    /// This expands to `f3(f2(f1(0)))` with the following `SyntaxContext`s:
+    ///
+    /// |Context     |Contents    |
+    /// |------------|------------|
+    /// |Root context|`f3(_)`, `0`|
+    /// |`m2!`       |`f2(_)`     |
+    /// |`m1!`       |`f1(_)`     |
+    ///
+    /// The following table lists the results of various possible argument combinations:
+    ///
+    /// |Span   |Context     |Result  |
+    /// |-------|------------|--------|
+    /// |`f3(_)`|Root        |`f3(_)` |
+    /// |`f3(_)`|`m1!`, `m2!`|None    |
+    /// |`f2(_)`|Root        |`m2!(0)`|
+    /// |`f2(_)`|`m2!`       |`f2(_)` |
+    /// |`f2(_)`|`m1!`       |None    |
+    /// |`f1(_)`|Root        |`m2!(0)`|
+    /// |`f1(_)`|`m2!`       |`m1!(0)`|
+    /// |`f1(_)`|`m1!`       |`f1(_)` |
+    /// |`0`    |Root        |`0`     |
+    /// |`0`    |`m1!`, `m2!`|None    |
+    #[inline]
+    #[must_use]
+    fn walk_to_ctxt(self, ctxt: SyntaxContext) -> Option<Span>
```

Comment:
This could do with some documentation on how it differs from `hygiene::walk_chain`